src/model.py

```diff
@@ -2083,44 +2033,42 @@ def init_components(self, subtasks):
                     out_channels=cnn_dim, 
                     kernel_size=kernel_size
                 ),
-                "attn_linear": self.linear_block(cnn_dim*2, 3)
             })
 
         for layer in range(1, shared_layers):
             # TODO use split_cnn_blocks in all following layers?
-            shared.update({
-                f"cnn_{layer}": self.cnn_block(
+            shared[f"cnn_{layer}"] = self.cnn_block(
                     in_channels=cnn_dim, 
                     out_channels=cnn_dim, 
                     kernel_size=kernel_size
                 )
-            })
-        components.update(torch.nn.ModuleDict({
-            "shared": shared
-        }))
-
-        # similar to RACL setup
-        relations = torch.nn.ModuleDict({
-            f"stack_{i}": torch.nn.ModuleDict({
-                f"{rel[0]}_at_{rel[1]}": self.attn_block(cnn_dim)
-                for rel in attention_relations
-            })
-            for i in range(stack_count)
-        })
-        components.update(torch.nn.ModuleDict({
-            "relations": relations,
-        }))
+        components["shared"] =  shared
 
-        if stack_count > 0:  # when stack == 0 no attention should be made. No task-wise cnn either.. 
-            prev_task = None
-            for task in self.subtasks:
-                for relation in components["relations"]["stack_0"]:
-                    if relation.startswith(task) and (task != prev_task):
-                        components[task]["attn_linear"] = self.linear_block(cnn_dim*2, 3)
-                        prev_task = task
+        ## Relation layers: similar to RACL setup
+        relations = torch.nn.ModuleDict({})
+        attn_linears = torch.nn.ModuleDict({})
+        for stack in range(stack_count):
+            for rel in attention_relations:
+                relations[f"stack_{stack}"] = torch.nn.ModuleDict({
+                    f"{rel[0]}_at_{rel[1]}": torch.nn.ModuleDict({
+                        "attn": self.attn_block(cnn_dim),
+
+                        # new linear for each attn relation for first_task logits using attn information
+                        "linear": self.linear_block(cnn_dim*2, 3),
+                    })
+                })
+                if rel[0] in self.subtasks + ["shared"] and "re_encode" not in components[rel[0]].keys():
+                    # only reach this one time per first_task=rel[0]
+                    count = sum([rel[0] == r[0] for r in attention_relations])
+                    # map subtask information back to shared hidden state size after (multiple) attention(s)
+                    components[rel[0]]["re_encode"] = self.linear_block(
+                        in_features=cnn_dim*(1 + count),  # 1 task output + number of relations with rel[0] as first task
+                        out_features=cnn_dim*2
+                    )
```

Comment:
A lot going on here.

First off, the if-statement:
We want to store the outputs of all the attentions a given subtask is a part of, before updating the next task input. The if checks that `first_task` is a subtask, and that it doesn't have a re-encoding layer form before.

`count`:
We get the number of times the current `first_task` occurs as a first task in `attention_relations`. This way, we know how many extra attn outputs will be concatenated.

Linear `re_encode` layer:
The input features to this layer need to account for this task's cnn output, plus all the times this task is a `first_task` in an attention. This way, we use all the information from every attention block during re-encoding for th next stack.